velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,356 @@
+/// In contrast, PrefixSortIterator returns char* point to the prefix data
+/// (not a "value_type" but a memory buffer store normalized keys) for memcmp
+/// and memcpy operators, means not ValueSwappable.
+/// This is also the reason we can not use std::sort directly, and implement
+/// quickSort in PrefixSortContext.
+struct PrefixSortIterator {
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix_(prefix), entrySize_(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
+      : prefix_(other.prefix_), entrySize_(other.entrySize_) {}
+
+  // Inline functions should be place in include file to avoid link error.
+  FOLLY_ALWAYS_INLINE char* operator*() const {
+    return prefix_;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator++() {
+    prefix_ += entrySize_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator--() {
+    prefix_ -= entrySize_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator++(int) {
+    auto tmp = *this;
+    prefix_ += entrySize_;
+    return tmp;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator--(int) {
+    auto tmp = *this;
+    prefix_ -= entrySize_;
+    return tmp;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator+(const uint64_t i) const {
+    auto result = *this;
+    result.prefix_ += i * entrySize_;
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator operator-(const uint64_t i) const {
+    PrefixSortIterator result = *this;
+    result.prefix_ -= i * entrySize_;
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE PrefixSortIterator& operator=(
+      const PrefixSortIterator& other) {
+    prefix_ = other.prefix_;
+    return *this;
+  }
+
+  FOLLY_ALWAYS_INLINE uint64_t
+  operator-(const PrefixSortIterator& other) const {
+    return (prefix_ - other.prefix_) / other.entrySize_;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator<(const PrefixSortIterator& other) const {
+    return prefix_ < other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator>(const PrefixSortIterator& other) const {
+    return prefix_ > other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator>=(const PrefixSortIterator& other) const {
+    return prefix_ >= other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator<=(const PrefixSortIterator& other) const {
+    return prefix_ <= other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator==(const PrefixSortIterator& other) const {
+    return prefix_ == other.prefix_;
+    ;
+  }
+
+  FOLLY_ALWAYS_INLINE bool operator!=(const PrefixSortIterator& other) const {
+    return prefix_ != other.prefix_;
+  }
+
+ private:
+  char* prefix_;
+  const uint64_t entrySize_;
+};
+} // namespace detail
+
+/// Provides a series of methods (mostly required by the sort
+/// algorithm) for PrefixSort. Maintains the buffer for swap operations and
+/// some necessary context information such as entrySize.
+class PrefixSortContext {
```

Comment:
```
class PrefixSort {
  void quickSort()
```

Reply:
@skadilover can we name PrefixSortContext to PrefixSortRunner? How about rename this to PrefixSortInternal.h

Reply:
1.  PrefixSortInternal.h : I think the 'Internal' is a bit broad concept, something like 'Encode/Decode' could also be a form of 'internal', howerver we only implement sort algrithm in this  header file.
2. PrefixSortRunner : I think 'Runner' may let reader relate it to someing with 'Thread',  but we have noting with thread at all, so I do not think it is a good idea.

Reply:
Context object usually provide environment variable kind of things but not execute operation directly. So it is better to find a better name for that.